lib/vrl/stdlib/src/insert.rs

```diff
@@ -0,0 +1,150 @@
+            },
+            Example {
+                title: "variable",
+                source: indoc! {r#"
+                    var = { "foo": true }
+                    insert!(value: var, path: ["bar"], data: "baz")
+                "#},
+                result: Ok(r#"{ "foo": true, "bar": "baz" }"#),
+            },
+            Example {
+                title: "invalid indexing",
+                source: r#"insert!(value: [], path: ["foo"], data: "baz")"#,
+                result: Ok(r#"{ "foo": "baz" }"#),
+            },
+            Example {
+                title: "invalid segment type",
+                source: r#"insert!({"foo": { "bar": [92, 42] }}, ["foo", true], "baz")"#,
+                result: Err(
+                    r#"function call error for "insert" at (0:59): path segment must be either "string" or "integer", not "boolean""#,
+                ),
+            },
+        ]
+    }
+
+    fn compile(
+        &self,
+        _state: &state::Compiler,
+        _ctx: &FunctionCompileContext,
+        mut arguments: ArgumentList,
+    ) -> Compiled {
+        let value = arguments.required("value");
+        let path = arguments.required("path");
+        let data = arguments.required("data");
+
+        Ok(Box::new(InsertFn { value, path, data }))
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct InsertFn {
+    value: Box<dyn Expression>,
+    path: Box<dyn Expression>,
+    data: Box<dyn Expression>,
+}
+
+impl Expression for InsertFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let path = match self.path.resolve(ctx)? {
+            Value::Array(segments) => {
+                let mut insert = LookupBuf::root();
+
+                for segment in segments {
+                    let segment = match segment {
+                        Value::Bytes(path) => {
+                            SegmentBuf::Field(String::from_utf8_lossy(&path).into_owned().into())
+                        }
+                        Value::Integer(index) => SegmentBuf::Index(index as isize),
+                        value => {
+                            return Err(format!(
+                                r#"path segment must be either "string" or "integer", not {}"#,
+                                value.kind()
+                            )
+                            .into())
+                        }
+                    };
+
+                    insert.push_back(segment)
+                }
+
+                insert
+            }
+            value => {
+                return Err(value::Error::Expected {
+                    got: value.kind(),
+                    expected: Kind::Array | Kind::Bytes,
+                }
+                .into())
+            }
+        };
+
+        let mut value = self.value.resolve(ctx)?;
+        value.insert(&path, self.data.resolve(ctx)?)?;
+
+        Ok(value)
+    }
+
+    fn type_def(&self, _: &state::Compiler) -> TypeDef {
+        TypeDef::new().fallible().unknown()
```

Comment:
The type returned by this function will be the same as the value parameter - either an array or an object. We should be able to specify that here?